File: hiroo-py/hiroo/voice.py
"""Voice connections for bots.

HiRoo delegates actual audio transport to LiveKit. A bot joins a voice channel
by:

  1. Asking the HiRoo API for a LiveKit token scoped to the target room.
  2. Connecting to the LiveKit SFU with the `livekit` Python SDK.

Because pulling in the ~30 MB livekit SDK would be painful for many users who
don't need voice, we only import it on demand. Install with:

    pip install livekit

Usage::

    vc = await bot.connect_voice(guild_id=..., channel_id=...)
    await vc.play_audio_file("greeting.opus")
    await vc.disconnect()
"""
from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass
from typing import Optional, TYPE_CHECKING

from .errors import HiRooError

if TYPE_CHECKING:
    from .client import Bot

log = logging.getLogger("hiroo.voice")


@dataclass
class VoiceConnection:
    bot: "Bot"
    room_name: str
    livekit_url: str
    livekit_token: str
    _room: object | None = None  # livekit.rtc.Room

    async def connect(self) -> None:
        try:
            from livekit import rtc
        except ImportError as e:
            raise HiRooError(
                "voice support requires the `livekit` package. Install with: pip install livekit"
            ) from e
        self._room = rtc.Room()
        await self._room.connect(self.livekit_url, self.livekit_token, rtc.RoomOptions(
            auto_subscribe=True, dynacast=True,
        ))
        log.info("connected to voice room %s", self.room_name)

    async def disconnect(self) -> None:
        if self._room is not None:
            try:
                await self._room.disconnect()
            finally:
                self._room = None

    async def play_audio_file(self, path: str) -> None:
        """Convenience: stream a local audio file. Requires ffmpeg + livekit."""
        try:
            from livekit import rtc
        except ImportError as e:
            raise HiRooError("livekit required") from e
        if self._room is None:
            raise HiRooError("not connected")
        # Decode via ffmpeg to 48kHz mono PCM16 and publish as a LiveKit audio track.
        source = rtc.AudioSource(48000, 1)
        track = rtc.LocalAudioTrack.create_audio_track("bot-audio", source)
        await self._room.local_participant.publish_track(track)

        proc = await asyncio.create_subprocess_exec(
            "ffmpeg", "-hide_banner", "-loglevel", "error",
            "-i", path, "-f", "s16le", "-ar", "48000", "-ac", "1", "-",
            stdout=asyncio.subprocess.PIPE,
        )
        assert proc.stdout is not None
        frame_ms = 10
        bytes_per_frame = (48000 // (1000 // frame_ms)) * 2  # 960 samples * 2 bytes
        try:
            while True:
                chunk = await proc.stdout.readexactly(bytes_per_frame)
                frame = rtc.AudioFrame.create(48000, 1, 480)
                frame.data[:] = chunk
                await source.capture_frame(frame)
        except asyncio.IncompleteReadError:
            pass
        finally:
            try: proc.kill()
            except Exception: pass
            await proc.wait()


async def _get_voice_token(bot: "Bot", room: str) -> tuple[str, str]:
    """Fetch a LiveKit token from the HiRoo API."""
    data = await bot.http.request("GET", "/api/voice/token", params={"room": room})
    return data["url"], data["token"]


async def connect_voice(bot: "Bot", *, guild_id: Optional[str] = None, channel_id: Optional[str] = None,
                        dm_id: Optional[str] = None) -> VoiceConnection:
    if channel_id:
        room = f"channel:{channel_id}"
    elif dm_id:
        room = f"dm:{dm_id}"
    else:
        raise ValueError("Provide channel_id or dm_id")
    url, token = await _get_voice_token(bot, room)
    vc = VoiceConnection(bot=bot, room_name=room, livekit_url=url, livekit_token=token)
    await vc.connect()
    return vc

File: hiroo-js/src/voice.js
// Голосовые подключения для ботов.
//
// HiRoo делегирует аудио-транспорт LiveKit. Бот подключается к voice-каналу
// так:
//   1. Запрашивает LiveKit-токен для конкретной комнаты через HiRoo API.
//   2. Коннектится к LiveKit-SFU через `@livekit/rtc-node`.
//
// Поскольку `@livekit/rtc-node` тянет ~80MB нативного бинаря, он объявлен
// как optionalDependencies — ставьте отдельно, если нужен голос:
//
//     npm install @livekit/rtc-node
//
// Применение:
//
//     const vc = await bot.connectVoice({ channelId: "..." });
//     await vc.playFile("./music.mp3");
//     await vc.disconnect();

import { spawn } from "node:child_process";
import { execSync } from "node:child_process";
import { HiRooError } from "./errors.js";

export class VoiceConnection {
  constructor(bot, roomName, url, token) {
    this.bot = bot;
    this.roomName = roomName;
    this.url = url;
    this.token = token;
    this._room = null;
    this._rtc = null;
  }

  async connect() {
    let rtc;
    try {
      rtc = await import("@livekit/rtc-node");
    } catch (e) {
      throw new HiRooError(
        "Voice support requires @livekit/rtc-node. Install with: npm install @livekit/rtc-node",
      );
    }
    this._rtc = rtc;
    this._room = new rtc.Room();
    await this._room.connect(this.url, this.token, {
      autoSubscribe: true,
      dynacast: true,
    });
    // Регистрируем voice-presence на HiRoo, чтобы фронт показал бота
    // в списке участников канала.
    try {
      await this.bot.http.request("POST", "/api/voice/presence", {
        json: { room: this.roomName },
      });
    } catch (e) {
      console.warn("[hiroo] voice presence register failed:", e?.message ?? e);
    }
  }

  async disconnect() {
    if (this._room) {
      try { await this._room.disconnect(); }
      finally { this._room = null; }
    }
    try { await this.bot.http.request("DELETE", "/api/voice/presence"); }
    catch {}
  }

  /**
   * Стримит локальный аудио-файл в активное подключение. Внутри поднимает
   * ffmpeg, декодирует в 48 kHz mono PCM16, публикует как LocalAudioTrack.
   * Требует ffmpeg в PATH.
   */
  async playFile(path) {
    if (!this._room) throw new HiRooError("not connected");
    const rtc = this._rtc;

    const ffmpeg = findFfmpeg();
    if (!ffmpeg) {
      throw new HiRooError(
        "ffmpeg не найден в PATH. Установите его:\n" +
        "  • Windows: winget install Gyan.FFmpeg  (или choco install ffmpeg)\n" +
        "  • macOS:   brew install ffmpeg\n" +
        "  • Linux:   apt install ffmpeg",
      );
    }

    // LiveKit: создаём источник, трек, публикуем с опцией microphone —
    // без неё SFU отвергает трек как не попадающий в token grants.
    const source = new rtc.AudioSource(48_000, 1);
    const track = rtc.LocalAudioTrack.createAudioTrack("bot-audio", source);
    const opts = new rtc.TrackPublishOptions();
    try { opts.source = rtc.TrackSource.SOURCE_MICROPHONE; } catch {}

    await withTimeout(
      this._room.localParticipant.publishTrack(track, opts),
      15_000,
      "publishTrack timed out — check that UDP 50000-50100 to the LiveKit server is reachable",
    );

    // 10 мс фреймы при 48 kHz mono → 480 сэмплов → 960 байт s16le.
    const samplesPerFrame = 480;
    const bytesPerFrame = samplesPerFrame * 2; // mono * 16-bit

    const proc = spawn(ffmpeg, [
      "-hide_banner", "-loglevel", "error",
      "-i", path,
      "-f", "s16le", "-ar", "48000", "-ac", "1", "-",
    ], { stdio: ["ignore", "pipe", "inherit"] });

    let buf = Buffer.alloc(0);
    try {
      for await (const chunk of proc.stdout) {
        buf = buf.length === 0 ? chunk : Buffer.concat([buf, chunk]);
        while (buf.length >= bytesPerFrame) {
          const slice = buf.subarray(0, bytesPerFrame);
          buf = buf.subarray(bytesPerFrame);
          // AudioFrame ожидает Int16Array. Копируем из Buffer.
          const int16 = new Int16Array(samplesPerFrame);
          for (let i = 0; i < samplesPerFrame; i++) {
            int16[i] = slice.readInt16LE(i * 2);
          }
          const frame = new rtc.AudioFrame(int16, 48_000, 1, samplesPerFrame);
          await source.captureFrame(frame);
        }
      }
    } finally {
      try { proc.kill(); } catch {}
    }
  }

  /** Заглушить локального участника (бота). */
  async setMuted(muted) {
    if (!this._room?.localParticipant) return;
    try { await this._room.localParticipant.setMicrophoneEnabled(!muted); }
    catch {}
  }
}

/** Получить URL и токен LiveKit для комнаты HiRoo. */
export async function getVoiceToken(bot, room) {
  const data = await bot.http.request("GET", "/api/voice/token", { params: { room } });
  return { url: data.url, token: data.token };
}

export async function connectVoice(bot, { channelId, dmId } = {}) {
  let room;
  if (channelId) room = `channel:${channelId}`;
  else if (dmId) room = `dm:${dmId}`;
  else throw new Error("Provide channelId or dmId");

  const { url, token } = await getVoiceToken(bot, room);
  const vc = new VoiceConnection(bot, room, url, token);
  await vc.connect();
  if (typeof bot._registerVoice === "function") bot._registerVoice(vc);
  return vc;
}

// ── internals ──────────────────────────────────────────────────────

function withTimeout(promise, ms, message) {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => reject(new HiRooError(message)), ms);
    promise.then(
      (v) => { clearTimeout(timer); resolve(v); },
      (e) => { clearTimeout(timer); reject(e); },
    );
  });
}

function findFfmpeg() {
  const cmd = process.platform === "win32" ? "where ffmpeg" : "command -v ffmpeg";
  try {
    return execSync(cmd, { stdio: ["ignore", "pipe", "ignore"] })
      .toString().trim().split(/\r?\n/)[0] || null;
  } catch { return null; }
}
